083-stretch-cluster.md

```diff
@@ -0,0 +1,155 @@
+metadata:
+  name: controller
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  replicas: 3
+  target:
+    clusterUrl: <K8S Cluster URL>
+    secret: <SecretName>
+  listenerConfig:
+    - configuration:
+          bootstrap:
+            alternativeNames:
+              - bootstrap-ext.<kubernetes infra host>
+              - broker-100.<kubernetes infra host>
+              - broker-101.<kubernetes infra host>
+            host: bootstrap-ext.<kubernetes infra host>
+          brokers:
+          - broker: 100
+            host: broker-100.<kubernetes infra host>
+          - broker: 101
+            host: broker-101.<kubernetes infra host>
+          class: nginx
+        name: connectext
+  roles:
+    - controller
+  storage:
+   .........
+```
+
+#### Kafka changes
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+  annotations:
+    strimzi.io/node-pools: enabled
+    strimzi.io/kraft: enabled
+    strimzi.io/stretch-mode: enabled
+spec:
+  kafka:
+    version: 3.7.0
+    metadataVersion: 3.7-IV4
+    listeners:
+      - port: 9093
+        tls: true
+        name: connectext
+        type: ingress
+```
+
+A new annotation (`stretch-mode: enabled`) will be introduced in Kafka custom resource to indicate when it is representing a stretch Kafka cluster. This approach is similar to how Strimzi currently enables features like KafkaNodePool (KNP) and KRaft mode.
+
+In a stretch Kafka cluster, we'll need bootstrap and broker services to be present on each Kubernetes cluster and be accessible from other clusters. The Kafka reconciler will identify all target clusters from KafkaNodePool resources and create these services in target Kubernetes clusters. This will ensure that even if the central cluster experiences an outage, external clients can still connect to the stretch cluster and continue their operations without interruption.
+
+#### Cross-cluster communication
+Kafka controllers/brokers are distributed across multiple Kubernetes environments and will need to communicate with each other. Currently, the Strimzi Kafka operator defines Kafka listeners for internal communication (controlplane and replication) between brokers/controllers (Kubernetes services using ports 9090 and 9091). The user is not able to influence how these services are set up and exposed outside the cluster. We would remove this limitation and allow users to define how these internal listeners are configured in the Kafka resource, just like they do for Kafka client listeners.
+
+Users will also be able to override listener configurations in each KafkaNodePool resource, if the listeners need to be exposed in different ways (ingress host names, Ingress annotations etc.) for each Kubernetes cluster. This will be similar to how KafkaNodePools are used to override other configuration like storage etc. To override a listener, KafkaNodePool will define configuration with same listner name as in Kafka resource.
```

Comment:
My understanding of this area is not as good so I might have misunderstood this but was thinking of scenarios where there might be a need for these to be different: host configuration for Ingress, dns, ingress related annotations etc.

Is there is a way to do define such variations from within a single Kafka resource? or can we put prerequisites that will mean on each cluster same configuration can be used? That would be much simpler.

Reply:
Well, I suggested it in the other threads -> I think having a clear prerequisites that the clusters support the same mechanisms makes sense to me. But as I also said, you need to think outside the box here. It is not about loadbalancers or Ingresses. It is (also) about things such as Submariner, Skupper, Istio Federation etc. that would overlay the Kubernetes clusters and abstract from us what exactly they use underneath.

Reply:
We did consider some of these and tried technologies like skupper and they do make networking much simpler as services on one cluster are visible on another cluster, but they do add a new dependency to the project and will need upfront setup from customers. So there are trade-offs.

I have briefly touched upon skupper in alternatives towards the bottom of the proposal, but I will elaborate more and we can reconsider if that is a cleaner approach that what is laid out here currently.

Reply:
But I think these things are what users are asking for - at least some of them. Because they already use them for other services. And you do not really want to have each piece of software use its own way to do things. And you likely also don't want to have ingresses or load balancers for every single project. So they need to be considered at least to the extent to make things extensible and be ready for them.

Reply:
I’ve been looking into technologies like `Skupper` and `Submariner` for cross-cluster communication. The idea is that Submariner/SKupper can help simplify communication between Kafka brokers and controllers across clusters by exposing services across k8s clusters, avoiding the need for complex external listeners. IMO Here’s how it could work

- Submariner/Skupper connects Kubernetes clusters, making it easier for Kafka brokers and controllers to communicate directly using their native IPs and DNS names across clusters. This is especially important for Kafka’s internal coordination and replication traffic.
- Submariner/Skupper can extend Kafka’s internal listeners (used for broker coordination and replication) to work across clusters, making the communication between brokers seamless without needing to configure external services like load balancers or Ingress.
- Submariner/Skupper uses encrypted tunnels for communication between clusters. Since Kafka already requires encrypted communication for its internal traffic, Submariner can enhance security without needing extra setup for mTLS between clusters.

## Set Up Submariner

- Install Submariner across the clusters where Kafka brokers are running.
- Ensure cross-cluster connectivity is established for the namespaces running Strimzi Kafka.

## Modify Kafka Listeners:
- Adjust the Kafka internal listeners to support cross-cluster communication through Submariner.

## Secure Communication
- Submariner’s encrypted tunnels will automatically secure communication, minimizing the need for additional mTLS configuration.

Reply:
Is the expectation to support multiple technologies for enabling cross-cluster communication?
For example:

- Kubernetes-native solutions like Ingresses and Load Balancers
- Skupper
- Submariner
- Istio Federation
- Linkerd Service Mesh
- Consul Connect
- Cilium
- ..............
- ..............
Will there be any preference for one solution over the others?

Reply:
As per the discussion here, we explored cross-cluster technologies such as Submariner and Skupper as potential solutions for achieving communication between Kafka brokers and controllers that are distributed across multiple Kubernetes environments. These tools facilitate cross-cluster communication by overlaying Kubernetes clusters without relying solely on traditional methods like LoadBalancers or Ingresses.

As part of our investigation, we manually deployed Kafka pods and necessary configuration, similar to what the Strimzi Kafka operator would normally do. The goal of these experiments was to evaluate the suitability, reliability, and overall feasibility of each technology, assessing their strengths and limitations for our use case.

To enable cross-cluster communication for a stretch Kafka cluster, the `advertised.listener` configuration needed to be adapted. In its default form, Strimzi creates headless services that support communication within a single Kubernetes cluster, using addresses such as **`my-cluster-broker-0.my-cluster-kafka-brokers.svc.cluster.local`**. However, these default service addresses are not accessible across cluster boundaries.


When multiple Kubernetes clusters were connected using Submariner, the broker/controller services of type clusterIp can be exported using Submariner to make them accessible by other clusters in the network.

By running the command **`subctl export service --kubeconfig <CONFIG> --namespace <NAMESPACE> my-cluster-kafka-brokers`**, we create a ServiceExport resource in the specified namespace. This resource signals Submariner to register the service with the Submariner Broker. The Broker acts as the coordinator for cross-cluster service discovery, leveraging the Lighthouse component to allow services in different clusters to find and communicate with each other. This process results in secure IP routing being configured between clusters. Submariner sets up tunnels and routing tables that enable direct traffic flow, overcoming the limitations of isolated cluster networks.


Once the service is exported, its fully qualified domain name becomes accessible as `<service-name>.<namespace>.svc.clusterset.local`. This global DNS name ensures that any cluster participating in the Submariner deployment can reach the service, facilitating the cross-cluster communication needed for Kafka brokers and controllers. For example, the `advertised.listener` configuration was updated from `my-cluster-broker-0.my-cluster-kafka-brokers.svc.cluster.local` to **`my-cluster-broker-0.cluster1.my-cluster-kafka-brokers.svc.clusterset.local`**, where **`cluster1`** represents the Submariner cluster ID. This update ensures that when a Kafka broker sends its advertised listeners to clients or other brokers, they will receive a service address that is reachable from any cluster involved in the setup.

Similar changes are required for controller.quorum.voters property as well.

SSL hostname verification between pods relies on SAN (Subject Alternative Name) entries in the certificates provided to the pods. For this verification to function in a stretched Kafka cluster using Submariner, the FQDNs (Fully Qualified Domain Names) of the Submariner-exported services need to be included in the pod certificates. This can be accomplished in two main ways:

The first method allows users to define the SANs for the brokers directly through the Kafka CR's listener configuration property. Users can input the Submariner-exported FQDNs in this field, which ensures the brokers inject these SANs into their certificates. For example, if there are two k8s clusters with four brokers (broker-0, broker-1, broker-100, broker-101), the listener configuration might look like this:

```yaml
listeners:
  - name: tls
    port: 9093
    type: internal
    tls: true
    configuration:
      bootstrap:
        alternativeNames:
          - my-cluster-broker-0.cluster2.my-cluster-kafka-brokers.strimzi.svc.clusterset.local
          - my-cluster-broker-1.cluster2.my-cluster-kafka-brokers.strimzi.svc.clusterset.local
          - my-cluster-broker-100.cluster2.my-cluster-kafka-brokers.strimzi.svc.clusterset.local
          - my-cluster-broker-101.cluster2.my-cluster-kafka-brokers.strimzi.svc.clusterset.local
 ```

Although this approach works, it injects the FQDN of every broker into every broker's certificate, which is not ideal.

Controller pods do not follow this approach because they do not inherit listener configurations from the Kafka CR. Instead, they use a single control plane listener (TCP 9090). To make this work for controller pods, users would need to configure the control plane listener in the CR to include the necessary SANs, Strimzi currently doesn't support this and this is not considered optimal .

A better approach is for the operator to automatically read the Submariner cluster ID from the CR (The CR should be extended such that user should be able to provide Submariner ClusterId) and create the SANs entries in these formats:

```
<KAFKA-CLUSTER-NAME>-kafka-brokers.svc.default.clusterset.local
<SUBMARINER-CLUSTER-ID>.<KAFKA-CLUSTER-NAME>-kafka-brokers.svc.default.clusterset.local
<POD-NAME>.<SUBMARINER-CLUSTER-ID>.<KAFKA-CLUSTER-NAME>-kafka-brokers.svc.default.clusterset.local
```


This second approach is preferred as it simplifies the process for users, sparing them from manually adding SANs to the CR and reducing configuration complexity.

In summary this is what will be needed 

### Advertised Listener Configuration
- The advertised.listener property should have Submariner exported Service 

for example advertised.listener should be updated 

From: **`my-cluster-broker-0.my-cluster-kafka-brokers.svc.cluster.local`**
To: **`my-cluster-broker-0.cluster1.my-cluster-kafka-brokers.svc.clusterset.local`**

### Controller Quorum Voters
- Similar changes needs to be made for the controller.quorum.voters setting, ensuring that it points to the Submariner-exposed service.

### SANs 
- All the Broker and Controller pods should have SANS entry for the Submariner exported service.

I will update the proposal in detail with all all these changes and will also add potential implementation details